File: ascii_maker/core/processor.py
"""Frame processing pipeline.

Resize → grayscale → brightness/contrast → dither/char-map → colorize.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field

import numpy as np
from PIL import Image

from ascii_maker.core.charsets import (
    CHARSETS,
    CharsetName,
    braille_from_array,
)
from ascii_maker.core.color import ColorMode, colorize_line
from ascii_maker.core.dither import floyd_steinberg
from ascii_maker.core.reader import Frame


@dataclass(frozen=True)
class Settings:
    """Processing settings that affect output."""

    charset: CharsetName = CharsetName.SIMPLE
    color_mode: ColorMode = ColorMode.TRUECOLOR
    dither: bool = False
    brightness: int = 0  # -100 to 100
    contrast: int = 100  # 0 to 200 (100 = no change)
    invert: bool = False
    width: int = 80
    height: int = 24

    def hash(self) -> str:
        """Deterministic hash for cache keying."""
        data = (
            f"{self.charset}:{self.color_mode}:{self.dither}:"
            f"{self.brightness}:{self.contrast}:{self.invert}:"
            f"{self.width}:{self.height}"
        )
        return hashlib.md5(data.encode()).hexdigest()[:12]


@dataclass
class ProcessedFrame:
    """Result of processing a single frame."""

    lines: list[str]  # Plain text lines (no color)
    colored_lines: list[str]  # Lines with ANSI color escapes
    duration_ms: int
    index: int
    width: int = 0
    height: int = 0


def _resize_frame(img: Image.Image, width: int, height: int) -> Image.Image:
    """Resize image to target character grid dimensions.

    Characters are roughly 1:2 aspect ratio (tall), so we compensate.
    """
    return img.resize((width, height), Image.Resampling.LANCZOS)


def _resize_for_braille(img: Image.Image, width: int, height: int) -> Image.Image:
    """Resize for braille: each braille char covers a 2x4 pixel grid."""
    pixel_w = width * 2
    pixel_h = height * 4
    return img.resize((pixel_w, pixel_h), Image.Resampling.LANCZOS)


def _to_grayscale(img: Image.Image) -> np.ndarray:
    """Convert to grayscale float array in [0.0, 1.0]."""
    return np.array(img.convert("L"), dtype=np.float64) / 255.0


def _adjust_brightness_contrast(
    gray: np.ndarray, brightness: int, contrast: int
) -> np.ndarray:
    """Apply brightness and contrast adjustments.

    brightness: -100 to 100 (added as offset)
    contrast: 0 to 200 (100 = no change, multiplied around midpoint)
    """
    result = gray.copy()

    # Brightness: shift
    if brightness != 0:
        result = result + brightness / 100.0

    # Contrast: scale around 0.5
    if contrast != 100:
        factor = contrast / 100.0
        result = (result - 0.5) * factor + 0.5

    return np.clip(result, 0.0, 1.0)


def _get_color_samples(
    img: Image.Image, width: int, height: int
) -> np.ndarray:
    """Get per-character RGB color by sampling center pixel of each cell.

    Returns array of shape (height, width, 3).
    """
    resized = img.resize((width, height), Image.Resampling.LANCZOS)
    return np.array(resized.convert("RGB"), dtype=np.uint8)


def _get_braille_color_samples(
    img: Image.Image, width: int, height: int
) -> np.ndarray:
    """Get per-braille-character RGB color by averaging each 2x4 block.

    Returns array of shape (height, width, 3).
    """
    pixel_w = width * 2
    pixel_h = height * 4
    resized = np.array(
        img.resize((pixel_w, pixel_h), Image.Resampling.LANCZOS).convert("RGB"),
        dtype=np.float64,
    )

    colors = np.zeros((height, width, 3), dtype=np.uint8)
    for y in range(height):
        for x in range(width):
            block = resized[y * 4 : (y + 1) * 4, x * 2 : (x + 1) * 2, :]
            colors[y, x] = block.mean(axis=(0, 1)).astype(np.uint8)
    return colors


def process_frame(frame: Frame, settings: Settings) -> ProcessedFrame:
    """Process a single frame through the full pipeline."""
    charset = CHARSETS[settings.charset]
    is_braille = charset.is_braille
    w, h = settings.width, settings.height

    # Resize
    if is_braille:
        resized = _resize_for_braille(frame.image, w, h)
    else:
        resized = _resize_frame(frame.image, w, h)

    # Grayscale
    gray = _to_grayscale(resized)

    # Brightness / contrast
    gray = _adjust_brightness_contrast(gray, settings.brightness, settings.contrast)

    # Invert
    if settings.invert:
        gray = 1.0 - gray

    # Dither or direct mapping
    if is_braille:
        # Braille: threshold to binary, optionally with dithering
        if settings.dither:
            gray = floyd_steinberg(gray, levels=2)
        else:
            gray = (gray > 0.5).astype(np.float64)
        binary = (gray > 0.5).astype(np.uint8)
        plain_lines = braille_from_array(binary)
    else:
        if settings.dither:
            gray = floyd_steinberg(gray, levels=len(charset.chars))
        plain_lines = charset.map_array(gray)

    # Color
    if settings.color_mode != ColorMode.NONE:
        if is_braille:
            color_samples = _get_braille_color_samples(frame.image, w, h)
        else:
            color_samples = _get_color_samples(frame.image, w, h)

        # Apply invert to color samples too if inverted
        if settings.invert:
            color_samples = 255 - color_samples

        colored = []
        for row_idx, line in enumerate(plain_lines):
            if row_idx < color_samples.shape[0]:
                row_colors = color_samples[row_idx, : len(line), :]
                colored.append(colorize_line(line, row_colors, settings.color_mode))
            else:
                colored.append(line)
    else:
        colored = list(plain_lines)

    return ProcessedFrame(
        lines=plain_lines,
        colored_lines=colored,
        duration_ms=frame.duration_ms,
        index=frame.index,
        width=w,
        height=h,
    )

File: web/js/processor.js
/**
 * Frame processing pipeline.
 * Resize → grayscale → brightness/contrast → dither/char-map → colorize.
 * Ported from ascii_maker/core/processor.py
 */

import { CHARSETS, mapArrayToChars, brailleFromArray } from "./charsets.js";
import { floydSteinberg } from "./dither.js";

/**
 * Resize ImageData to target dimensions using canvas.
 * @param {ImageData} imageData
 * @param {number} width
 * @param {number} height
 * @returns {ImageData}
 */
function resizeImageData(imageData, width, height) {
  // Create source canvas
  const srcCanvas = document.createElement("canvas");
  srcCanvas.width = imageData.width;
  srcCanvas.height = imageData.height;
  const srcCtx = srcCanvas.getContext("2d");
  srcCtx.putImageData(imageData, 0, 0);

  // Create dest canvas and draw scaled
  const dstCanvas = document.createElement("canvas");
  dstCanvas.width = width;
  dstCanvas.height = height;
  const dstCtx = dstCanvas.getContext("2d");
  dstCtx.imageSmoothingEnabled = true;
  dstCtx.imageSmoothingQuality = "high";
  dstCtx.drawImage(srcCanvas, 0, 0, width, height);

  return dstCtx.getImageData(0, 0, width, height);
}

/**
 * Convert ImageData to grayscale float array.
 * @param {ImageData} imageData
 * @returns {Float32Array[]} - 2D array of rows, values [0, 1]
 */
function toGrayscale(imageData) {
  const { width, height, data } = imageData;
  const rows = [];

  for (let y = 0; y < height; y++) {
    const row = new Float32Array(width);
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * 4;
      // Standard luminance formula
      const lum = (0.299 * data[i] + 0.587 * data[i + 1] + 0.114 * data[i + 2]) / 255;
      row[x] = lum;
    }
    rows.push(row);
  }
  return rows;
}

/**
 * Apply brightness and contrast adjustments.
 * @param {Float32Array[]} gray
 * @param {number} brightness - -100 to 100
 * @param {number} contrast - 0 to 200 (100 = no change)
 * @returns {Float32Array[]}
 */
function adjustBrightnessContrast(gray, brightness, contrast) {
  const h = gray.length;
  const w = gray[0]?.length || 0;
  const result = gray.map((row) => new Float32Array(row));

  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      let val = result[y][x];

      // Brightness: shift
      if (brightness !== 0) {
        val = val + brightness / 100.0;
      }

      // Contrast: scale around 0.5
      if (contrast !== 100) {
        const factor = contrast / 100.0;
        val = (val - 0.5) * factor + 0.5;
      }

      result[y][x] = Math.max(0.0, Math.min(1.0, val));
    }
  }
  return result;
}

/**
 * Get per-character RGB colors by sampling resized image.
 * @param {ImageData} imageData - original image
 * @param {number} width - target width
 * @param {number} height - target height
 * @returns {Uint8Array[][]} - 2D array of [r, g, b] values
 */
function getColorSamples(imageData, width, height) {
  const resized = resizeImageData(imageData, width, height);
  const { data } = resized;
  const colors = [];

  for (let y = 0; y < height; y++) {
    const row = [];
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * 4;
      row.push([data[i], data[i + 1], data[i + 2]]);
    }
    colors.push(row);
  }
  return colors;
}

/**
 * Get per-braille-character RGB colors by averaging 2x4 blocks.
 * @param {ImageData} imageData - original image
 * @param {number} width - character width
 * @param {number} height - character height
 * @returns {Uint8Array[][]}
 */
function getBrailleColorSamples(imageData, width, height) {
  const pixelW = width * 2;
  const pixelH = height * 4;
  const resized = resizeImageData(imageData, pixelW, pixelH);
  const { data } = resized;
  const colors = [];

  for (let y = 0; y < height; y++) {
    const row = [];
    for (let x = 0; x < width; x++) {
      let rSum = 0,
        gSum = 0,
        bSum = 0;
      const count = 8; // 2x4 block

      for (let dy = 0; dy < 4; dy++) {
        for (let dx = 0; dx < 2; dx++) {
          const px = x * 2 + dx;
          const py = y * 4 + dy;
          const i = (py * pixelW + px) * 4;
          rSum += data[i];
          gSum += data[i + 1];
          bSum += data[i + 2];
        }
      }

      row.push([
        Math.round(rSum / count),
        Math.round(gSum / count),
        Math.round(bSum / count),
      ]);
    }
    colors.push(row);
  }
  return colors;
}

/**
 * Process a frame through the full pipeline.
 * @param {ImageData} imageData - source frame
 * @param {Object} settings
 * @param {string} settings.charset - 'simple', 'detailed', 'blocks', 'braille'
 * @param {string} settings.colorMode - 'none', 'truecolor'
 * @param {boolean} settings.dither
 * @param {number} settings.brightness - -100 to 100
 * @param {number} settings.contrast - 0 to 200
 * @param {boolean} settings.invert
 * @param {number} settings.width
 * @param {number} settings.height
 * @returns {Object} - { lines, colors, width, height }
 */
export function processFrame(imageData, settings) {
  const {
    charset,
    colorMode,
    dither,
    brightness,
    contrast,
    invert,
    width,
    height,
  } = settings;

  const isBraille = charset === "braille";

  // Resize
  let resizedWidth, resizedHeight;
  if (isBraille) {
    resizedWidth = width * 2;
    resizedHeight = height * 4;
  } else {
    resizedWidth = width;
    resizedHeight = height;
  }

  const resized = resizeImageData(imageData, resizedWidth, resizedHeight);

  // Grayscale
  let gray = toGrayscale(resized);

  // Brightness / contrast
  gray = adjustBrightnessContrast(gray, brightness, contrast);

  // Invert
  if (invert) {
    for (let y = 0; y < gray.length; y++) {
      for (let x = 0; x < gray[y].length; x++) {
        gray[y][x] = 1.0 - gray[y][x];
      }
    }
  }

  let plainLines;

  if (isBraille) {
    // Braille: threshold to binary, optionally with dithering
    if (dither) {
      gray = floydSteinberg(gray, 2);
    }
    // Convert to binary
    const binary = gray.map((row) => {
      const binRow = new Uint8Array(row.length);
      for (let x = 0; x < row.length; x++) {
        binRow[x] = row[x] > 0.5 ? 1 : 0;
      }
      return binRow;
    });
    plainLines = brailleFromArray(binary);
  } else {
    // Normal charsets
    if (dither) {
      gray = floydSteinberg(gray, CHARSETS[charset].length);
    }
    plainLines = mapArrayToChars(charset, gray);
  }

  // Color samples
  let colorSamples = null;
  if (colorMode !== "none") {
    if (isBraille) {
      colorSamples = getBrailleColorSamples(imageData, width, height);
    } else {
      colorSamples = getColorSamples(imageData, width, height);
    }

    // Apply invert to colors too
    if (invert) {
      for (let y = 0; y < colorSamples.length; y++) {
        for (let x = 0; x < colorSamples[y].length; x++) {
          colorSamples[y][x] = [
            255 - colorSamples[y][x][0],
            255 - colorSamples[y][x][1],
            255 - colorSamples[y][x][2],
          ];
        }
      }
    }
  }

  return {
    lines: plainLines,
    colors: colorSamples,
    width,
    height,
  };
}
